Add pybind type caster for c10::Device

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57354 Inline hooks in ivalue::Future
* #57353 Use Devices instead of DeviceIndexes in Future
* #57294 Use Devices instead of DeviceIndexes in TensorPipe agent
* #57345 Rename DeviceGuardImplInteface's getStreamFromPool method
* #57293 Add device() method to c10::Event
* **#57292 Add pybind type caster for c10::Device**

In Future (and soon in other places too) we need to receive a list of devices from Python-land. We don't want to just take their indices because we need full devices in order to infer the type from them. torch.device is not defined through pybind, it's defined through a plain `PyModule_AddObject` call with CPython, thus pybind isn't naturally able to understand and convert it. However we can provide a custom type caster which fixes that. We have this already for at::Tensor, at::Generator, ...

Differential Revision: [D28092732](https://our.internmc.facebook.com/intern/diff/D28092732/)

lw committed:
Add pybind type caster for c10::Device

In Future (and soon in other places too) we need to receive a list of devices from Python-land. We don't want to just take their indices because we need full devices in order to infer the type from them. torch.device is not defined through pybind, it's defined through a plain `PyModule_AddObject` call with CPython, thus pybind isn't naturally able to understand and convert it. However we can provide a custom type caster which fixes that. We have this already for at::Tensor, at::Generator, ...

Differential Revision: [D28092732](https://our.internmc.facebook.com/intern/diff/D28092732/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/utils/pybind.h:
```diff
@@ -102,6 +103,33 @@ template<> struct type_caster<at::IntArrayRef> {
   std::vector<int64_t> v_value;
 };
 
+template <>
```
This is great! It gonna save a lot of code in both Python and C++. Thanks!

lw committed:
Update on "Add pybind type caster for c10::Device"


In Future (and soon in other places too) we need to receive a list of devices from Python-land. We don't want to just take their indices because we need full devices in order to infer the type from them. torch.device is not defined through pybind, it's defined through a plain `PyModule_AddObject` call with CPython, thus pybind isn't naturally able to understand and convert it. However we can provide a custom type caster which fixes that. We have this already for at::Tensor, at::Generator, ...

Differential Revision: [D28092732](https://our.internmc.facebook.com/intern/diff/D28092732/)

[ghstack-poisoned]

lw committed:
Update on "Add pybind type caster for c10::Device"


In Future (and soon in other places too) we need to receive a list of devices from Python-land. We don't want to just take their indices because we need full devices in order to infer the type from them. torch.device is not defined through pybind, it's defined through a plain `PyModule_AddObject` call with CPython, thus pybind isn't naturally able to understand and convert it. However we can provide a custom type caster which fixes that. We have this already for at::Tensor, at::Generator, ...

Differential Revision: [D28092732](https://our.internmc.facebook.com/intern/diff/D28092732/)

[ghstack-poisoned]

lw committed:
Update on "Add pybind type caster for c10::Device"


In Future (and soon in other places too) we need to receive a list of devices from Python-land. We don't want to just take their indices because we need full devices in order to infer the type from them. torch.device is not defined through pybind, it's defined through a plain `PyModule_AddObject` call with CPython, thus pybind isn't naturally able to understand and convert it. However we can provide a custom type caster which fixes that. We have this already for at::Tensor, at::Generator, ...

Differential Revision: [D28092732](https://our.internmc.facebook.com/intern/diff/D28092732/)

[ghstack-poisoned]
